### DISCUSSION TITLE
invisible clone "ghost submarine" in subeditor test mode. ?

### DISCUSSION THREAD

**Discussion by Ek-Videogames** [FakeFishGames/Barotrauma#14590] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

Some background: 
_"In subeditor test mode i've sometimes put the sub's engines on full power to stress test the power grid, or to test some other submarine feature. Only to suddenly collide with some kind of large object, originally whenever this happened I just assumed this was some invisible wall at the 'edge' of the world bounds but on further investigation the truth is something much stranger."_

There is an invisible, seemingly indestructible, and frozen in place clone of whatever submarine you are testing floating out in the testing map void. (or possibly multiple?) In this reproduction it seems to be floating at exactly x 50000 y 15000 which seems significant. The dimensions of the submarine does not seem to affect the "ghost ship's" relative position?

I tried turning on debugdraw to see if maybe the invisible structure hitboxes would actually show up, but they didn't. Interestingly though the ragdoll points of the crewmembers on my ship were being 'copied' by the ghost ship - this is a potentially useful detail for locating the "ghost ship" in reproducing this bug in the future.

My submarine seen parked physically on top of the 'ghost ship'
![image](https://github.com/user-attachments/assets/106da762-95e9-41a7-93f1-dd4a6839e6be)

Here I reproduced the result with the Dugong for reference. I don't know what those floating green lines above it are...
![image](https://github.com/user-attachments/assets/0d26dbb5-8b29-4c7f-86da-103d2faa4fb5)

Projectiles will collide with the "ghost ship's" walls, but don't do any damage. Strangely, you can damage your own DOORS this way however! Hitscan projectiles seem to behave this way as well. (In fact I think I may have at some point in the past killed my own submarine with a random stray railgun shell because I fired it into the abyss and hit my own 'ghost' but I didn't know that at the time.)
![image](https://github.com/user-attachments/assets/c86218bd-775c-4f3c-9091-dfcd63d643fa)
![image](https://github.com/user-attachments/assets/343efb79-60e0-43a4-a8bb-462af7acb85e)

Characters collide with the "ghost ship" the same way the submarine does... funny enough, if you teleportcharacter yourself inside you can swim around inside the invisible interior walls of the ghost ship.
![image](https://github.com/user-attachments/assets/9ea53dbe-f8d5-452e-af24-f6e69c7d6988)

You can shoot your ghost crewmembers too!
![image](https://github.com/user-attachments/assets/e59fc35d-d9df-4d74-b403-1b8e2b01ed3c)

Also it's odd this hasn't been reported before, since it seems like it may have existed for a while and not unreasonably hard to stumble upon, since it's not like it's a thousand kilometers away from spawn or anything only a few hundred meters. Its highly likely someone else has experienced this and just thought nothing of it at the time.

### Reproduction steps

- Start subeditor, load a sub, enter test mode.
- (optional) Enable debugdraw to make the "ghost" crew ragdolls visible.
- Drive to x 50000 y 15000 (or freecam if you want to).
- Comedy ensues.

### Bug prevalence

Happens regularly

### Single player or multiplayer?

Other

### Version

v1.5.9.2 (Summer Update Hotfix 2)

### -

I have previously observed this bug happen on unstable 1.6.4 build, although I'm reproducing the bug for this report on the current stable.

### Which operating system did you encounter this bug on?

Windows

### Other Information

This is such a obscure and specific "low priority" bug unlikely to really affect anything in real gameplay, but I guess if anyone feels like they want an absolute head-scratcher of a ticket well here it is...

**Comment by SomeRandomNoobKekeke** [comment]
This is actually normal game behaviour :BaroDev:
https://github.com/FakeFishGames/Barotrauma/blob/0e8fb6569d2810e2f8ad5fb17b4bba546cc5739a/Barotrauma/BarotraumaShared/SharedSource/Map/Submarine.cs#L29

It's some physics optimisation: 
Submarines have invisible outer body that is just intersection of hulls and walls, it's physically moving around the level and colliding with walls
While everything that is inside the sub is located in some hidden static position above level top and just drawn like it's on the level

All impact impulses are fake and whenever something enters the sub it gets teleported to hidden position
It's also stops colliding with level walls so [devs kindly spawn some invisible plank under your butt so you don't clip inside walls](https://github.com/FakeFishGames/Barotrauma/issues/13525)
And that's why your camera seems teleporting when you move from one sub to another in multiplayer
And that's where [all the scooters](https://github.com/FakeFishGames/Barotrauma/issues/7606#issuecomment-2002523058) and [diving suits](https://github.com/FakeFishGames/Barotrauma/discussions/14112#discussioncomment-9734541) went to

But in sub editor there's no top level wall so you can actually find it

**Comment by Ek-Videogames** [comment]
Yeah I'm aware of how the 'instancing' of the physics works for the submarine hull, it just kind of slipped my mind that the origin point for submarine physics simulation is so near the "roof" of a normal level that you can reach it in subeditor like that. One of those things you just sorta forget when it's not campaign mode.

I think the last time I really though about it was back in legacy & the early versions of the steam beta when you used to be able to actually bits of artifacting from the submarine hulls 'stored' on top of the map... I feel like I have the vaguest memory that they used to be stored directly 'above' the start of the map, and it wasn't until later they were moved towards the 'middle' instead but that might be me talking out of my ass.

Seems... 'modular' enough? Even though it doesn't really matter - **worse case you could use just moving it like another 500 meters upwards when in subeditor test**, maybe part of the reason why it's so easy to run into accidentally is because when a submarine first spawns in (the 'automatic' buoyancy detection isn't 100% accurate especially when dealing with unusual hull shapes or ballast setups) the submarine usually drifts upwards for a few seconds until the power kicks on & the nav terminal and the pumps stabilize it, so driving straight forwards after floating upwards for a few seconds ends up giving a chance to get really unlucky and ram right into it.
